[Documentation] Quick reference guide for rocprofv3

bgopesh opened the pull request:
## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

bgopesh committed:
quick reference guide for rocprofv3

subbul-amd commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,323 @@
+   rocprofv3 --runtime-trace --output-format csv -- ./your_app
+
+   # System-level tracing (includes HSA API)
+   rocprofv3 --sys-trace --output-format json -- ./your_app
```
for consistency can you keep csv format in all examples or json ( or in comment you can mention either of them as option)

subbul-amd commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,323 @@
+   rocprofv3 --pmc SQ_WAVES,SQ_INSTS_VALU --output-format csv -- ./your_app
+
+   # List available counters
+   rocprofv3-avail list --pmc
```
please move this as first command in "Performance Counter collection" as it could be easy to discover all counters

bgopesh replied:
done

subbul-amd commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,323 @@
+   rocprofv3 --pc-sampling-beta-enabled --pc-sampling-interval 1000 --output-format csv -- ./your_app
+
+   # Check PC sampling support
+   rocprofv3-avail list --pc-sampling
```
Check for support of PC sampling first - before using, so can you please move it as first example (in this section)

subbul-amd commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,323 @@
+Output Formats and Post-processing
+===================================
+
+Output Format Options
```
Default (rocpd) can be captured?

bgopesh replied:
yes

subbul-amd commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,323 @@
+.. code-block:: bash
+
+   # Trace ROCTx markers and ranges
+   rocprofv3 --marker-trace --output-format csv -- ./your_app
```
repeated in "Granular Tracing Options"

bgopesh replied:
done

bgopesh committed:
Addressed feedback and updated with rocpd information

bgopesh committed:
rocpd docs update

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +200,256 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
 
    /opt/rocm/bin/rocpd convert -i db{3,4}.db --output-format csv otf2 pftrace
 
+Dedicated Conversion Tools
+++++++++++++++++++++++++++
```
In addition to adding this section, I was hoping we add a "Summary" section, and perhaps a section to highlight "Aggregating rocpd data".  You may also want to add a "Query" section since that is another module we added now too.  Like, when they do `rocpd --help`, they'll see {convert, query, summary}.

Summary: basically is providing the same summaries that rocprofv3 used to provide, but using the DB data.  So you can link to https://rocm.docs.amd.com/projects/rocprofiler-sdk/en/latest/how-to/using-rocprofv3.html#summary, to explain.

Aggregation: We should call out that rocpd allows multiple databases (runs on different ranks/nodes) can be combined together for analysis, something we couldn't do with rocprofv3 perfetto output before.  

===
Also, rocpd Summary has an added benefit, in that, it can aggregate multiple DB data, and we want to highlight that.  If we have multiple databases, you can summarize them together, using the built in summary reports.  BUT if you want to analyze them by rank, you can use `--summary-by-rank` to get an analysis comparing one rank/node to another.

bgopesh replied:
Done

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +200,256 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+
+ROCprofiler-SDK provides specialized conversion utilities for efficient format-specific operations. These tools offer streamlined interfaces for single-format conversions and are particularly useful in automated workflows and scripts.
+
+rocpd2csv - CSV Export Tool
```
I didn't think we needed to go into so much detail for each of these "direct conversion" scripts, but I guess we can keep them for now.

bgopesh replied:
Sure, I think we should keep them as well for now

gsitaram commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,415 @@
+   # OTF2 format (performance analysis)
+   rocprofv3 --runtime-trace --output-format otf2 -- ./your_app
+
+**Documentation:** :ref:`using-rocpd-output-format`
```
I looked through the changes in this PR but did not see a deprecation notice for the `--output-format` option for `rocprofv3` that is expected to go away in ROCm 7.1. If this is the plan, could that be added?

bgopesh replied:
I will talk to Jonathan/Young

yhuiYH replied:
We should double-check what Jonathan has planned.  I know, initially he wanted to drop the "legacy outputs" in ROCm 7.1, but I think he may delay that deprecation.

bgopesh committed:
rocpd query option

bgopesh committed:
Addressing feedback

bgopesh committed:
Fixed misssing newline

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+.. code-block:: bash
+
+   # Include all available domains
+   rocpd2summary -i profile.db --region-categories HIP ROCR MARKERS KERNEL
+
+   # Focus on GPU kernel analysis only
+   rocpd2summary -i profile.db --region-categories KERNEL
+
+   # Exclude markers to speed up processing
+   rocpd2summary -i profile.db --region-categories HIP ROCR KERNEL
```
```suggestion
   rocpd2summary -i profile.db --region-categories HIP HSA MARKERS KERNEL

   # Focus on GPU kernel analysis only
   rocpd2summary -i profile.db --region-categories KERNEL

   # Exclude markers to speed up processing
   rocpd2summary -i profile.db --region-categories HIP HSA KERNEL
```
Do we actually have a ROCR API category?  When I do sys-trace, I usually just get: HIP HSA MARKER categories only.

bgopesh replied:
No! you are right. I have removed the `ROCr` mention.

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+**Output Files:**
+- ``summary_kernel.{format}`` - GPU kernel execution summary
+- ``summary_hip_api.{format}`` - HIP API call statistics
+- ``summary_rocr_api.{format}`` - ROCr runtime API analysis
```
I see you have ROCr here too, maybe the workload you ran has ROCr api traces.  If so, you can ignore my above comment.

bgopesh replied:
No! you are right. I have removed the ROCr mention.

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+   
+   -- API trace information
+   SELECT * FROM hip_api_trace;
+   SELECT * FROM rocr_api_trace;
```
```suggestion
   SELECT * FROM regions_and_samples WHERE category LIKE 'HIP_%'
   SELECT * FROM regions_and_samples WHERE category LIKE 'RCCL_%'
```
We don't actually have the _API_TRACE views anymore.  They need a slightly more complicated query to get that data, like that ^

bgopesh replied:
done

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+   
+   -- Marker and region data
+   SELECT * FROM regions;
+   SELECT * FROM markers;
```
```suggestion
   SELECT * FROM regions_and_samples WHERE category LIKE 'MARKERS_%'
```
Markers comes from a query like this

bgopesh replied:
done

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+   SELECT * FROM api_call_statistics;
+   
+   -- Device utilization metrics
+   SELECT * FROM device_utilization;
```
These views/tables don't exist so those won't work.  We should remove these
```suggestion
 
```

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+
+   # Kernel performance analysis with statistics
+   rocpd query -i profile.db --query "
+   SELECT 
+       kernel_name,
+       COUNT(*) as dispatch_count,
+       MIN(duration_ns) as min_duration,
+       AVG(duration_ns) as avg_duration,
+       MAX(duration_ns) as max_duration,
+       SUM(duration_ns) as total_duration
+   FROM kernels 
+   GROUP BY kernel_name 
+   ORDER BY total_duration DESC"
```
```suggestion
   SELECT 
       name,
       COUNT(*) as dispatch_count,
       MIN(duration) as min_duration,
       AVG(duration) as avg_duration,
       MAX(duration) as max_duration,
       SUM(duration) as total_duration
   FROM kernels 
   GROUP BY name 
   ORDER BY total_duration DESC"
```
Fixed the fields to match the kernel view column names.  This works, I ran the query.

bgopesh replied:
Thanks! Done

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+
+.. code-block:: bash
+
+   # Memory copy performance by direction
+   rocpd query -i profile.db --query "
+   SELECT 
+       copy_kind,
+       COUNT(*) as transfer_count,
+       SUM(size_bytes) as total_bytes,
+       AVG(bandwidth_gb_per_sec) as avg_bandwidth
+   FROM memory_copies 
+   GROUP BY copy_kind 
+   ORDER BY total_bytes DESC"
```
```suggestion
   # Memory copy analysis by direction
   rocpd query -i profile.db --query "
   SELECT 
       name,
       src_agent_type,
       src_agent_abs_index,
       dst_agent_type,
       dst_agent_abs_index,
       COUNT(*) as transfer_count,
       SUM(size) as total_bytes,
       SUM(duration) as total_duration
   FROM memory_copies 
   GROUP BY src_agent_abs_index 
   ORDER BY total_bytes DESC"
```
Changed this query to make it work.  Had to drop the avg_bandwidth because that's not stored by default.

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+
+   -- Create temporary views for complex analysis
+   CREATE TEMP VIEW kernel_stats AS
+   SELECT 
+       kernel_name,
+       COUNT(*) as dispatch_count,
+       AVG(duration_ns) as avg_duration,
+       STDDEV(duration_ns) as duration_stddev
+   FROM kernels 
+   GROUP BY kernel_name;
+   
+   CREATE TEMP VIEW performance_outliers AS
+   SELECT k.*, ks.avg_duration, ks.duration_stddev
+   FROM kernels k
+   JOIN kernel_stats ks ON k.kernel_name = ks.kernel_name
+   WHERE ABS(k.duration_ns - ks.avg_duration) > 2 * ks.duration_stddev;
```
```suggestion
   SELECT 
       name,
       COUNT(*) as dispatch_count,
       AVG(duration) as avg_duration,
       STDDEV(duration) as duration_stddev
   FROM kernels 
   GROUP BY name;
   
   CREATE TEMP VIEW performance_outliers AS
   SELECT k.*, ks.avg_duration, ks.duration_stddev
   FROM kernels k
   JOIN kernel_stats ks ON k.name = ks.name
   WHERE ABS(k.duration - ks.avg_duration) > 2 * ks.duration_stddev;
```

Fixed the query to use proper column names

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+- **Automated Reporting:** Email delivery with SMTP configuration and attachment management
+- **Time Window Integration:** Apply temporal filtering before query execution
+
+Database Schema and Views
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+rocpd databases provide comprehensive views for analysis:
+
```
```suggestion
Database Schema and Views
~~~~~~~~~~~~~~~~~~~~~~~~~

rocpd databases provide comprehensive views for analysis.  In general, any queries should be built using the `data_views`:

```
We want people to just use the data_views, not any of the other tables.  Maybe we point them to where the schema files are located in our repository so anyone interested can view the .sql files, to help them build queries?

https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/source/share/rocprofiler-sdk-rocpd/data_views.sql

bgopesh replied:
done

yhuiYH reviewed (commented):
Overall good content.  I fixed a bunch of queries because they seem to be generated as general queries, instead of using our actual SQL views.  I would like them fixed because some ppl may copy/paste them to try to use them, and will fail.  We should search the .rst file for any mentions of `kernel_name` for example, to see if there are other spots that need to be fixed.

bgopesh committed:
Addressing feedback

bgopesh commented:
> Overall good content. I fixed a bunch of queries because they seem to be generated as general queries, instead of using our actual SQL views. I would like them fixed because some ppl may copy/paste them to try to use them, and will fail. We should search the .rst file for any mentions of `kernel_name` for example, to see if there are other spots that need to be fixed.

I checked all the places and replaced `kernel_name` with `name` and `duration_ns` with `duration`

bgopesh commented:
Hi @yhuiYH : Thanks for the review. Have addressed all.

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+   
+   # Aggregate device utilization analysis
+   rocpd query -i multi_gpu_results.db \
+               --query "SELECT device_id, COUNT(*) as operations, SUM(duration) as total_time FROM kernels GROUP BY device_id"
```
```suggestion
               --query "SELECT agent_abs_index as device_id, COUNT(*) as operations, SUM(duration) as total_time FROM kernels GROUP BY device_id"
```
Found another one, can make it readable like this.

bgopesh replied:
Yes, makes sense

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+   -- Top performing kernels by execution time
+   SELECT * FROM top_kernels LIMIT 10;
+   
+   -- Memory bandwidth analysis
+   SELECT * FROM memory_copy_summary;
```
```suggestion
   -- Top analysis
   SELECT * FROM top;
   
   -- Busy analysis
   SELECT * FROM busy;
```
We also don't have memory_copy_summary.  All summary queries are dynamically generated in python (as requested by JM).  He doesn't want any summary views stored in the schema.  So all we have stored, are top, top_kernels, busy.

bgopesh replied:
Got it

yhuiYH commented on projects/rocprofiler-sdk/source/docs/how-to/using-rocpd-output-format.rst:
```diff
@@ -200,3 +216,883 @@ Convert multiple databases to all supported formats (CSV, OTF2, and Perfetto tra
+.. code-block:: bash
+
+   # Kernel performance analysis with statistics
+   rocpd query -i profile.db --query "
+   SELECT 
+       name,
+       COUNT(*) as dispatch_count,
+       MIN(duration) as min_duration,
+       AVG(duration) as avg_duration,
+       MAX(duration) as max_duration,
+       SUM(duration) as total_duration
+   FROM kernels 
+   GROUP BY name 
+   ORDER BY total_duration DESC"
```
```suggestion
   rocpd query -i profile.db --query "
   SELECT 
       name as kernel_name,
       COUNT(*) as dispatch_count,
       MIN(duration) as min_duration,
       AVG(duration) as avg_duration,
       MAX(duration) as max_duration,
       SUM(duration) as total_duration
   FROM kernels 
   GROUP BY kernel_name
   ORDER BY total_duration DESC"
```
Just realized, we can do this, to make it more readable... like you had it before.  If you like it this way, then, feel free to make the changes to the rest of the queries.

Otherwise, just ignore this comment and just leave them.  Up to you.

bgopesh replied:
yes, its more readable

bgopesh committed:
Addressing feedback

yhuiYH reviewed (approved):
LGTM

jrmadsen commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,415 @@
+
+.. code-block:: bash
+
+   export PATH=$PATH:/opt/rocm/bin
```
Use `source /opt/rocm/share/rocprofiler-sdk/setup-env.sh`

bgopesh replied:
Thanks! wasn't aware on this.

jrmadsen commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,415 @@
+   rocprofv3 --runtime-trace --output-format rocpd -- ./your_app
+
+   # Query the database directly with SQL
+   sqlite3 hostname/12345_results.db "SELECT * FROM rocprofiler_domain_info;"
```
This is an invalid rocpd table. I'd suggest `SELECT * FROM regions;`

bgopesh replied:
done

jrmadsen commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,415 @@
+   import rocpd
+   rocpd.rocpd_to_csv('hostname/12345_results.db', 'output.csv')
+   rocpd.rocpd_to_json('hostname/12345_results.db', 'output.json')
+   "
```
Use the rocpd command line tools. Also, these are not valid function calls in rocpd

bgopesh replied:
done.

jrmadsen commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,415 @@
+
+ROCprofiler-SDK provides dedicated conversion utilities for transforming rocpd databases to different formats:
+
+**rocpd2csv - CSV Export:**
```
There is a generic rocpd command line tool which has a subparser "convert"

bgopesh replied:
That has been mentioned under `Using rocpd Output Format.rst` This is only quick-reference, so didn't want to mention everything here.

jrmadsen replied:
The quick-start guide should just use `rocpd` since it has all of the functionality. And you only need to give a 1-2 examples of each subparser. This is fairly heavy for a quick start guide

bgopesh replied:
ok, sure

bgopesh replied:
I have either removed lots of extra information OR moved them in a single line. Also removed all the --output-format stuff except when used with ATT.

jrmadsen commented on projects/rocprofiler-sdk/source/docs/quick_guide.rst:
```diff
@@ -0,0 +1,415 @@
+.. code-block:: bash
+
+   # Generate timeline for visualization tools
+   rocprofv3 --runtime-trace --output-format pftrace -- ./your_app
```
Recommend rocpd

bgopesh replied:
done.

jrmadsen reviewed (changes requested):
There are significant issues with rocpd content.

bgopesh committed:
Addressing feedback

jrmadsen reviewed (changes requested):
Move all the CSV, Perfetto, OTF2 into a "conversion", e.g. `rocpd convert -i *.db -f csv pftrace otf2 --start 20% --end %80` demonstrates plenty. If they want more details, point them to the other page.

bgopesh committed:
Addressing feedback

bgopesh commented:
> rocpd convert -i *.db -f csv pftrace otf2 --start 20% --end %80

Done

bgopesh commented:
> Move all the CSV, Perfetto, OTF2 into a "conversion", e.g. `rocpd convert -i *.db -f csv pftrace otf2 --start 20% --end %80` demonstrates plenty. If they want more details, point them to the other page.

Done

bgopesh commented:
@jrmadsen Addressed feedback

bgopesh committed:
Adding process attachment
